WalletWasabi/WabiSabi/Backend/Banning/CoinVerifier.cs

```diff
@@ -21,87 +22,95 @@ public CoinVerifier(CoinJoinIdStore coinJoinIdStore, CoinVerifierApiClient apiCl
-	{
-		CoinJoinIdStore = coinJoinIdStore;
-		CoinVerifierApiClient = apiClient;
-		Whitelist = new();
-		WabiSabiConfig = wabiSabiConfig;
-	}
-
-	public Whitelist Whitelist { get; }
-	public WabiSabiConfig WabiSabiConfig { get; }
+	private Whitelist Whitelist { get; }
+	private WabiSabiConfig WabiSabiConfig { get; }
 	private CoinJoinIdStore CoinJoinIdStore { get; }
 	private CoinVerifierApiClient CoinVerifierApiClient { get; }
+	private ConcurrentDictionary<Coin, (DateTimeOffset ScheduleTime, TaskCompletionSource<CoinVerifyResult> TaskCompletionSource)> CoinVerifyItems { get; } = new();
 
-	private bool CheckIfAlreadyVerified(Coin coin)
-	{
-		// Step 1: Check if address is whitelisted.
-		if (Whitelist.TryGet(coin.Outpoint, out _))
-		{
-			return true;
-		}
-
-		// Step 2: Check if address is from a coinjoin.
-		if (CoinJoinIdStore.Contains(coin.Outpoint.Hash))
-		{
-			return true;
-		}
-
-		return false;
-	}
+	public event EventHandler<Coin>? CoinBlackListed;
 
-	public async IAsyncEnumerable<CoinVerifyInfo> VerifyCoinsAsync(IEnumerable<Coin> coinsToCheck, [EnumeratorCancellation] CancellationToken cancellationToken, string roundId = "")
+	public async Task<IEnumerable<CoinVerifyResult>> VerifyCoinsAsync(IEnumerable<Coin> coinsToCheck, CancellationToken cancellationToken)
 	{
-		var before = DateTimeOffset.UtcNow;
-
 		using CancellationTokenSource cancellationTokenSource = new(TimeSpan.FromSeconds(30));
 		using CancellationTokenSource linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationTokenSource.Token, cancellationToken);
 
-		var lastChangeId = Whitelist.ChangeId;
-		Whitelist.RemoveAllExpired(WabiSabiConfig.ReleaseFromWhitelistAfter);
-
-		var scriptsToCheck = new HashSet<Script>();
-		var innocentsCounter = 0;
+		// Booting up the results with the default value - ban: no, remove: yes.
+		Dictionary<Coin, CoinVerifyResult> coinVerifyItems = coinsToCheck.ToDictionary(c => c, c => new CoinVerifyResult(c, false, true));
 
+		// Building up the task list.
+		List<Task<CoinVerifyResult>> tasks = new();
 		foreach (var coin in coinsToCheck)
 		{
-			if (CheckIfAlreadyVerified(coin))
-			{
-				innocentsCounter++;
-				yield return new CoinVerifyInfo(false, coin);
-			}
-			else
+			if (!CoinVerifyItems.TryGetValue(coin, out var item))
 			{
-				scriptsToCheck.Add(coin.ScriptPubKey);
+				// If the coin was not scheduled try to quickly schedule it - it should not happen.
+				Logger.LogWarning($"Trying to re-schedule coin '{coin.Outpoint}' for verification.");
+
+				// Quickly re-scheduling the missing items.
+				ScheduleVerification(coin, cancellationToken, TimeSpan.Zero);
+				if (!CoinVerifyItems.TryGetValue(coin, out item))
+				{
+					// This should not happen.
+					Logger.LogError($"Coin '{coin.Outpoint}' cannot be re-scheduled for verification.");
+				}
 			}
+
+			tasks.Add(item.TaskCompletionSource.Task);
```

Comment:
If we tried to add the Coin 2 times already, and still didn't succeed, we should put a `continute` here or something, as in that case `item` very likely will be `null`, so `tasks.Add(item.TaskCompletionSource.Task)` will cause us trouble.

Reply:
What if: 

```cs
foreach (var coin in coinsToCheck)
{
  if (!CoinVerifyItems.TryGetValue(coin, out var item))
  {
	  // If the coin was not scheduled try to quickly schedule it - it should not happen.
	  Logger.LogWarning($"Trying to re-schedule coin '{coin.Outpoint}' for verification.");
  
	  // Quickly re-scheduling the missing items.
	  ScheduleVerification(coin, cancellationToken, TimeSpan.Zero);
	  if (!CoinVerifyItems.TryGetValue(coin, out item))
	  {
		  // This should not happen.
		  Logger.LogError($"Coin '{coin.Outpoint}' cannot be re-scheduled for verification.");
		  continue;
	  }
  }
}
List<Task<CoinVerifyResult>> tasks = CoinVerifyItems.Values.Select(x => x.TaskCompletionSource.Task).ToList();
```

Reply:
fixed